### ISSUE TITLE
Can't specify memory space for internal Bitset in UnorderedMap construction

### ISSUE THREAD

**Issue by rppawlo** [kokkos/kokkos#3475]
When we create a  Kokkos::UnorderedMap, an internal Bitset is created using the execution space of the UnorderedMap. This means that the Bitset allocation uses the default memory space nested in that execution space. This is normally not an issue unless the execution space is Cuda and the default memory space is configured to be UVM (in Trilinos). We are trying to make all of Trilinos libraries UVM free but if applications are counting on UVM and using the Default space on Cuda, they might configure UVM default on. Can the Bitset and UnorderedMap be modified to not use the default space for allocation?

As an example, I hacked the memory events profile tool to show the allocation spaces for a phalanx test where trilinos is configured with UVM enabled. You can see that the Bitset in the UnorderedMap is still using UVM. If I reconfigure Trilinos with UVM disabled then everything is allocated in non-UVM CudaSpace.
```
SpaceHandle=Host, allocation name=grad_basis_ref_mirror
SpaceHandle=Cuda, allocation name=workset_det_jac
SpaceHandle=Cuda, allocation name=workset_grad_basis_real
SpaceHandle=Cuda, allocation name=workset_det_jac
SpaceHandle=Cuda, allocation name=workset_grad_basis_real
SpaceHandle=Cuda, allocation name=workset_det_jac
SpaceHandle=Cuda, allocation name=workset_grad_basis_real
SpaceHandle=Cuda, allocation name=workset_det_jac
SpaceHandle=Cuda, allocation name=workset_grad_basis_real
SpaceHandle=Cuda, allocation name=workset_det_jac
SpaceHandle=Cuda, allocation name=workset_grad_basis_real
SpaceHandle=Cuda, allocation name=workset_det_jac
SpaceHandle=Cuda, allocation name=workset_grad_basis_real
SpaceHandle=CudaUVM, allocation name=Bitset
SpaceHandle=Cuda, allocation name=UnorderedMap hash list
SpaceHandle=Cuda, allocation name=UnorderedMap next index
SpaceHandle=Cuda, allocation name=UnorderedMap keys
SpaceHandle=Cuda, allocation name=UnorderedMap values
SpaceHandle=Cuda, allocation name=UnorderedMap scalars
SpaceHandle=Cuda, allocation name=row_count
SpaceHandle=Cuda, allocation name=row_offsets
SpaceHandle=Cuda, allocation name=matrix_size
SpaceHandle=CudaUVM, allocation name=Bitset
SpaceHandle=Cuda, allocation name=UnorderedMap hash list
SpaceHandle=Cuda, allocation name=UnorderedMap next index
SpaceHandle=Cuda, allocation name=UnorderedMap keys
SpaceHandle=Cuda, allocation name=UnorderedMap values
SpaceHandle=Cuda, allocation name=UnorderedMap scalars
SpaceHandle=Cuda, allocation name=graph_entries
```

Note - the phalanx test code above was refactored to not use the default mem space anymore. This has not been pushed into Trilinos yet, but should be added soon. 

Also - I realize you can still use UVM with the configure flag for UVM disabled. The configure flag only toggles the default memory space. Unfortunately many applications rely on this functionality.

**Comment by crtrott** [comment]
Yeah we need to check whether there is an actual reason for the BitSet to be UVM here. There probably is though (my guess is related to the underlying memory pool).

**Comment by masterleinad** [comment]
Fixed in #4675.

**Comment by PhilMiller** [comment]
Fixed in 3.6
